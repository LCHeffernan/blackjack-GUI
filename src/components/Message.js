import React, { useEffect, useState } from "react";
import PropTypes from "prop-types";

const Message = ({ playerHand, dealerHand }) => {
  const [message, setMessage] = useState("Do you want to play?");

  useEffect(() => {
    if (!playerHand.isHandValid) {
      setMessage("BUST - GAME OVER!");
    } else if (
      playerHand.playerScore === 21 &&
      playerHand.playerHand.length === 2
    ) {
      if (dealerHand.dealerScore === 21 && dealerHand.dealerHand.length === 2) {
        setMessage("DRAW - You both have blackjack");
      } else {
        setMessage("♥️ ♣️ BLACKJACK ♦️ ♠️");
      }
    } else if (playerHand.isGameOver && playerHand.isHandValid) {
      if (dealerHand.dealerScore > 21) {
        setMessage("Dealer bust, you WIN");
      } else if (dealerHand.dealerScore >= 17 && dealerHand.dealerScore <= 21) {
        if (dealerHand.dealerScore > playerHand.playerScore) {
          setMessage("Sorry you LOSE");
        } else if (dealerHand.dealerScore < playerHand.playerScore) {
          setMessage("Congratulations you WIN");
        } else {
          setMessage("DRAW");
        }
      }
    } else {
      setMessage("Enjoy your game");
    }
  }, [playerHand, dealerHand]);

  return <div className="message">{message}</div>;
};

Message.defaultProps = {
  playerHand: {},
  dealerHand: {},
};

Message.propTypes = {
  playerHand: PropTypes.shape({
    playerHand: PropTypes.arrayOf(
      PropTypes.shape({
        cardRank: PropTypes.string,
        cardSuit: PropTypes.string,
        cardValue: PropTypes.number,
      })
    ),
    playerScore: PropTypes.number,
    isHandValid: PropTypes.bool,
    isGameOver: PropTypes.bool,
  }),
  dealerHand: PropTypes.shape({
    dealerHand: PropTypes.arrayOf(
      PropTypes.shape({
        cardRank: PropTypes.string,
        cardSuit: PropTypes.string,
        cardValue: PropTypes.number,
      })
    ),
    dealerScore: PropTypes.number,
    isHandValid: PropTypes.bool,
    isGameOver: PropTypes.bool,
  }),
};

export default Message;
