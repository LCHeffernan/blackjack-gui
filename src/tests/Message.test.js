import React, { render, screen } from "@testing-library/react";
import Message from "../components/Message";

describe("Message", () => {
  const validProps = {
    playerHand: {
      playerHand: [
        {
          cardRank: "Mock",
          cardSuit: "Mock",
          cardValue: 0,
        },
        {
          cardRank: "Mock",
          cardSuit: "Mock",
          cardValue: 0,
        },
      ],
      playerScore: 0,
      isHandValid: true,
      isGameOver: false,
    },
    dealerHand: {
      dealerHand: [
        {
          cardRank: "Mock",
          cardSuit: "Mock",
          cardValue: 0,
        },
        {
          cardRank: "Mock",
          cardSuit: "Mock",
          cardValue: 0,
        },
      ],
      dealerScore: 0,
      isHandValid: true,
      isGameOver: false,
    },
  };

  it("renders correctly", () => {
    const { asFragment } = render(
      <Message
        playerHand={validProps.playerHand}
        dealerHand={validProps.dealerHand}
      />
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it("renders correct message upon game started", () => {
    render(
      <Message
        playerHand={validProps.playerHand}
        dealerHand={validProps.dealerHand}
      />
    );

    expect(screen.getByText("Enjoy your game")).toHaveClass("message-text");
  });

  it("renders correct message upon player having bust hand", () => {
    validProps.playerHand.isHandValid = false;
    render(
      <Message
        playerHand={validProps.playerHand}
        dealerHand={validProps.dealerHand}
      />
    );

    expect(screen.getByText("BUST - GAME OVER!")).toHaveClass("message-text");
  });

  it("renders correct message if player and dealer have blackjack", () => {
    validProps.playerHand.isHandValid = true;
    validProps.playerHand.playerScore = 21;
    validProps.dealerHand.dealerScore = 21;

    render(
      <Message
        playerHand={validProps.playerHand}
        dealerHand={validProps.dealerHand}
      />
    );

    expect(screen.getByText("DRAW - You both have blackjack")).toHaveClass(
      "message-text"
    );
  });

  it("renders correct message if player has blackjack", () => {
    validProps.dealerHand.dealerScore = 20;

    render(
      <Message
        playerHand={validProps.playerHand}
        dealerHand={validProps.dealerHand}
      />
    );

    expect(screen.getByText("♥️ ♣️ BLACKJACK ♦️ ♠️")).toHaveClass("message-text");
  });

  it("renders correct message if dealer is bust", () => {
    validProps.playerHand.isGameOver = true;
    validProps.playerHand.playerScore = 18;
    validProps.dealerHand.dealerScore = 22;

    render(
      <Message
        playerHand={validProps.playerHand}
        dealerHand={validProps.dealerHand}
      />
    );

    expect(screen.getByText("Dealer bust, you WIN")).toHaveClass("message-text");
  });

  it("renders correct message if dealer has higher score than player ", () => {
    validProps.dealerHand.dealerScore = 20;

    render(
      <Message
        playerHand={validProps.playerHand}
        dealerHand={validProps.dealerHand}
      />
    );

    expect(screen.getByText("Sorry you LOSE")).toHaveClass("message-text");
  });

  it("renders correct message if player has higher score than dealer ", () => {
    validProps.dealerHand.dealerScore = 17;

    render(
      <Message
        playerHand={validProps.playerHand}
        dealerHand={validProps.dealerHand}
      />
    );

    expect(screen.getByText("Congratulations you WIN")).toHaveClass("message-text");
  });

  it("renders correct message if player has same score as dealer (not blackjack) ", () => {
    validProps.dealerHand.dealerScore = 18;

    render(
      <Message
        playerHand={validProps.playerHand}
        dealerHand={validProps.dealerHand}
      />
    );

    expect(screen.getByText("DRAW")).toHaveClass("message-text");
  });
});
