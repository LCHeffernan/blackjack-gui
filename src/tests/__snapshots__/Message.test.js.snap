// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Message renders correctly 1`] = `
<DocumentFragment>
  <div
    class="message-text"
  >
    Enjoy your game
  </div>
</DocumentFragment>
`;
